DuckDuckGo/SwipeTabsCoordinator.swift

```diff
@@ -0,0 +1,359 @@
+        } else {
+            collectionView.horizontalScrollIndicatorInsets.bottom = collectionView.frame.height - 7.5
+            collectionView.hitTestInsets.top = 0
+            collectionView.hitTestInsets.bottom = -12
+        }
+    }
+    
+}
+
+// MARK: UICollectionViewDataSource
+extension SwipeTabsCoordinator: UICollectionViewDataSource {
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard isEnabled else { return 1 }
+        let extras = tabsModel.tabs.last?.link != nil ? 1 : 0 // last tab is not a home page, so let's add one
+        let count = tabsModel.count + extras
+        return count
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "omnibar", for: indexPath) as? OmniBarCell else {
+            fatalError("Not \(OmniBarCell.self)")
+        }
+
+        if !isEnabled || tabsModel.currentIndex == indexPath.row {
+            cell.omniBar = coordinator.omniBar
+        } else {
+            cell.omniBar = OmniBar.loadFromXib()
+            cell.omniBar?.translatesAutoresizingMaskIntoConstraints = false
+            cell.updateConstraints()
+            cell.omniBar?.decorate(with: ThemeManager.shared.currentTheme)
+            
+            cell.omniBar?.showSeparator()
+            if self.appSettings.currentAddressBarPosition.isBottom {
+                cell.omniBar?.moveSeparatorToTop()
+            } else {
+                cell.omniBar?.moveSeparatorToBottom()
+            }
+
+            if let url = tabsModel.safeGetTabAt(indexPath.row)?.link?.url {
+                cell.omniBar?.startBrowsing()
+                cell.omniBar?.refreshText(forUrl: url)
+            }
+
+        }
+        
+        return cell
+    }
+    
+}
+
+class OmniBarCell: UICollectionViewCell {
+    
+    weak var omniBar: OmniBar? {
+        didSet {
+            subviews.forEach { $0.removeFromSuperview() }
+            if let omniBar {
+                addSubview(omniBar)
+                
+                NSLayoutConstraint.activate([
+                    constrainView(omniBar, by: .leadingMargin),
+                    constrainView(omniBar, by: .trailingMargin),
+                    constrainView(omniBar, by: .top),
+                    constrainView(omniBar, by: .bottom),
+                ])
+                
+            }
+        }
+    }
+    
+    override func updateConstraints() {
+        super.updateConstraints()
+        let left = superview?.safeAreaInsets.left ?? 0
+        let right = superview?.safeAreaInsets.right ?? 0
+        omniBar?.updateOmniBarPadding(left: left, right: right)
+    }
+    
+}
+
+extension TabsModel {
+    
+    func safeGetTabAt(_ index: Int) -> Tab? {
+        guard tabs.indices.contains(index) else { return nil }
+        return tabs[index]
```

Comment:
Would it make sense to have an assertionFailure in here so that we can catch when it comes up? Or is this expected to happen in legitimate usage?

Reply:
I'm assuming I can get a nil tab here tbh.